cluster-agent/utils/argocd_create.go

```diff
@@ -0,0 +1,453 @@
+			return fmt.Errorf("Error on DELETE %v", errOnDelete)
+		}
+
+	}
+	if err := k8sClient.Create(context.Background(), &clusterRole); err != nil {
+		if !apierr.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create clusterRole %q in namespace %q: %v", clusterRole.Name, clusterRole.Namespace, err)
+		}
+		return fmt.Errorf("clusterRole %q already exists in namespace %q", clusterRole.Name, clusterRole.Namespace)
+	}
+	log.Printf("clusterRole %q created in namespace %q", clusterRole.Name, clusterRole.Namespace)
+
+	clusterRoleBinding := rbac.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "argocd-manager-role-binding",
+			Namespace: "kube-system",
+		},
+		RoleRef: rbac.RoleRef{
+			APIGroup: "rbac.authorization.k8s.io",
+			Kind:     "ClusterRole",
+			Name:     "argocd-manager-role",
+		},
+		Subjects: []rbac.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      "argocd-manager",
+				Namespace: "kube-system",
+			},
+		},
+	}
+
+	// debug : cluster role binding doesn't respond to the below code, need to delete manually using kubectl.
+	_, errOnGet = kubeClientSet.RbacV1().ClusterRoleBindings().Get(context.Background(), clusterRoleBinding.Name, metav1.GetOptions{})
+	if errOnGet == nil {
+		errOnDelete := kubeClientSet.RbacV1().ClusterRoleBindings().Delete(context.Background(), clusterRoleBinding.Name, metav1.DeleteOptions{PropagationPolicy: &policy})
+		if errOnDelete != nil {
+			return fmt.Errorf("Error on DELETE %v", errOnDelete)
+		}
+	}
+	if err := k8sClient.Create(context.Background(), &clusterRoleBinding); err != nil {
+		if !apierr.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create clusterRoleBinding %q in namespace %q: %v", clusterRoleBinding.Name, clusterRoleBinding.Namespace, err)
+		}
+		return fmt.Errorf("clusterRoleBinding %q already exists in namespace %q", clusterRoleBinding.Name, clusterRoleBinding.Namespace)
+	}
+	log.Printf("clusterRoleBinding %q created in namespace %q", clusterRoleBinding.Name, clusterRoleBinding.Namespace)
+
+	// 	apiVersion: v1
+	// kind: Secret
+	// metadata:
+	//   name: my-cluster-secret
+	//   labels:
+	//     argocd.argoproj.io/secret-type: cluster
+	// type: Opaque
+	// stringData:
+	//   name: my-cluster
+	//   server: $API_URL
+	//   config: |
+	//     {
+	//       "bearerToken": "$SECRET_TOKEN",
+	//       "tlsClientConfig": {
+	//         "insecure": true
+	//       }
+	//     }
+
+	secretResource, secretErr := kubeClientSet.CoreV1().Secrets(secret.Namespace).Get(context.Background(), secret.Name, metav1.GetOptions{})
+	if secretErr != nil {
+		return secretErr
+	}
+	// token := secret.Data["token"]  not sure whether this will work
+	token := secretResource.Data["token"]
+	// no need to decode token, it is unmarshalled from base64
+
+	// decodedToken, err := base64.StdEncoding.DecodeString(string(token))
+	// if err != nil {
+	// 	fmt.Printf("Error decoding Base64 encoded data %v", err)
+	// }
+
+	type ClusterSecretTLSClientConfigJSON struct {
```

Comment:
TIL that you can define types within Go functions :smile:. This is useful information, but I think we can just move these type structs out of the function, for now.